src/coprocessor/codec/mysql/duration.rs

```diff
@@ -571,23 +714,179 @@ mod tests {
         for (lhs, rhs, exp) in cases.clone() {
             let lhs = Duration::parse(lhs.as_bytes(), 6).unwrap();
             let rhs = Duration::parse(rhs.as_bytes(), 6).unwrap();
-            let res = lhs.checked_add(&rhs).unwrap();
+            let res = lhs.checked_add(rhs).unwrap();
             let exp = Duration::parse(exp.as_bytes(), 6).unwrap();
             assert_eq!(res, exp);
         }
         for (exp, rhs, lhs) in cases {
             let lhs = Duration::parse(lhs.as_bytes(), 6).unwrap();
             let rhs = Duration::parse(rhs.as_bytes(), 6).unwrap();
-            let res = lhs.checked_sub(&rhs).unwrap();
+            let res = lhs.checked_sub(rhs).unwrap();
             let exp = Duration::parse(exp.as_bytes(), 6).unwrap();
             assert_eq!(res, exp);
         }
 
         let lhs = Duration::parse(b"00:00:01", 6).unwrap();
-        let rhs = Duration::from_nanos(MAX_TIME_IN_SECS as i64 * NANOS_PER_SEC, 6).unwrap();
-        assert_eq!(lhs.checked_add(&rhs), None);
+        let rhs = Duration::from_nanos((MAX_TIME_IN_SECS * NANOS_PER_SEC) as i64, 6).unwrap();
+        assert_eq!(lhs.checked_add(rhs), None);
         let lhs = Duration::parse(b"-00:00:01", 6).unwrap();
-        let rhs = Duration::from_nanos(MAX_TIME_IN_SECS as i64 * NANOS_PER_SEC, 6).unwrap();
-        assert_eq!(lhs.checked_sub(&rhs), None);
+        let rhs = Duration::from_nanos((MAX_TIME_IN_SECS * NANOS_PER_SEC) as i64, 6).unwrap();
+        assert_eq!(lhs.checked_sub(rhs), None);
+    }
+}
+
+#[cfg(test)]
+mod benches {
+    use super::*;
+    use crate::coprocessor::codec::mysql::MAX_FSP;
+    #[bench]
+    fn bench_parse(b: &mut test::Bencher) {
+        let cases = vec![
+            ("12:34:56.1234", 0),
+            ("12:34:56.789", 1),
+            ("1:2:3.189", 2),
+            ("2 27:54:32.828", 3),
+            ("2 33:44:55.666777", 4),
+            ("112233.445566", 5),
+            ("1 23", 5),
+            ("1 23:12.1234567", 6),
+        ];
+        b.iter(|| {
+            let cases = test::black_box(&cases);
+            for _ in 0..1000 {
+                for &(s, fsp) in cases {
+                    let _ = test::black_box(Duration::parse(s.as_bytes(), fsp).unwrap());
+                }
+            }
+        })
+    }
+
+    #[bench]
+    fn bench_hours(b: &mut test::Bencher) {
+        let cases: Vec<_> = vec![
+            ("12:34:56.1234", 0i8),
+            ("12:34:56.789", 1),
+            ("1:2:3.189", 2),
+            ("2 27:54:32.828", 3),
+            ("2 33:44:55.666777", 4),
+            ("112233.445566", 5),
+            ("1 23", 5),
+            ("1 23:12.1234567", 6),
+        ]
+        .into_iter()
+        .map(|(s, fsp)| Duration::parse(s.as_bytes(), fsp).unwrap())
+        .collect();
+        b.iter(|| {
+            let cases = test::black_box(&cases);
+            for _ in 0..1000 {
+                for duration in cases {
+                    let _ = test::black_box(duration.hours());
```

Comment:
- `for _ in ...` is not necessary.
- For hours, a single case should be enough because different cases do not affect its performance.